worker/agent.go

```diff
@@ -32,143 +39,216 @@ func newAgent(net, addr string, worker *Worker) (a *agent, err error) {
 	return
 }
 
-func (a *agent) Connect() (err error) {
-	a.Lock()
-	defer a.Unlock()
-	a.conn, err = net.Dial(a.net, a.addr)
-	if err != nil {
-		return
-	}
-	a.rw = bufio.NewReadWriter(bufio.NewReader(a.conn),
-		bufio.NewWriter(a.conn))
-	go a.work()
-	return
-}
-
 func (a *agent) work() {
+	a.worker.Log(Info, "Starting agent Work For:", a.addr)
 	defer func() {
 		if err := recover(); err != nil {
-			a.worker.err(err.(error))
+			a.reconnect_error(err.(error))
 		}
 	}()
 
 	var inpack *inPack
 	var l int
 	var err error
 	var data, leftdata []byte
-	for {
-		if !a.worker.isShuttingDown() {
-			if data, err = a.read(); err != nil {
-				if opErr, ok := err.(*net.OpError); ok {
-					if opErr.Temporary() {
-						continue
-					} else {
-						a.disconnect_error(err)
-						// else - we're probably dc'ing due to a Close()
-
-						break
-					}
-
-				} else if err == io.EOF {
-					a.disconnect_error(err)
-					break
-				}
-				a.worker.err(err)
-				// If it is unexpected error and the connection wasn't
-				// closed by Gearmand, the agent should close the conection
-				// and reconnect to job server.
-				a.Close()
-				a.conn, err = net.Dial(a.net, a.addr)
-				if err != nil {
-					a.worker.err(err)
+	startConnVersion := a.connectionVersion
+
+	// exit the loop if connection has been replaced because reconnect will launch a new work() thread
+	for startConnVersion == a.connectionVersion && !a.worker.isShuttingDown() {
```

Comment:
Why not just look at the [pointer `a.conn`](https://play.golang.org/p/lw-kjSIbP4-) points to instead of keeping this `connectionVersion` value around?